fastlane_bot/helpers/wrap_unwrap_processor.py

```diff
@@ -0,0 +1,344 @@
+from typing import Any, List, Dict
+
+from fastlane_bot.helpers import TradeInstruction, RouteStruct
+from dataclasses import asdict
+
+
+class WrapUnwrapProcessor:
+    """
+    Handles the decision-making for wrapping & unwrapping gas tokens within trade routes.
+
+    This class tracks token balances of trades, and decides when to add wrap or unwrap trades.
+
+    Typical usage example:
+    wrap_unwrap_processor = WrapUnwrapProcessor(cfg=self.ConfigObj)
+    route_struct_processed = wrap_unwrap_processor.add_wrap_or_unwrap_trades_to_route(trade_instructions=split_calculated_trade_instructions, route_struct=route_struct, flashloan_struct=flashloan_struct)
+
+    """
+
+    def __init__(self, cfg: Any):
+        self.cfg = cfg
+
+    def add_wrap_or_unwrap_trades_to_route(
+        self,
+        trade_instructions: List[TradeInstruction],
+        route_struct: List[Dict],
+        flashloan_struct: List,
+    ) -> List[Dict]:
+        """Adds wrap or unwrap trade actions to a route struct.
+
+        This method takes in trade instructions, route struct and flashloan struct and returns a new route struct with wrap/unwrap trades added.
+
+        Args:
+            trade_instructions: a list of TradeInstruction objects
+            route_struct: a list of Dictionaries containing the route struct to be submitted to the Fast Lane contract.
+            flashloan_struct: a list containing flashloan instructions
+
+        Returns:
+            The route struct - an ordered List of Dictionaries - with the added wrap/unwrap trades.
+
+        """
+        self._init_balance_tracker(flashloan_struct)
+        self._ensure_unique_flashloan_token()
+        segmented_routes = self._segment_routes(trade_instructions, route_struct)
+        return self._generate_new_route_struct(segmented_routes, route_struct)
```

Comment:
Generally speaking, it is pointless to implement a non-api (internal) function which is used only once within its module.
And there are 4 such functions here.
It hides out potential improvements within what is technically a single block of execution.
Only good reason to do so, is when these functions are large, which isn't the case here.

Reply:
Splitting it out into functions made it much easier to test the individual processes - previously this was one single giant function.